### PR TITLE
WIP: Coach Reports MultiSelect

### PR DESCRIPTION
**Pull request by sethachoi**
WIP: addition of multi-topic select for coach reports

Default view
screen shot 2015-11-23 at 2 16 08 pm
Topic selection
screen shot 2015-11-23 at 2 16 21 pm
Topics selected
screen shot 2015-11-23 at 2 18 27 pm

### CODE REVIEW

**Comment by sethachoi** [comment]
Two bugs (I know of) currently: 
1. The total exercise attempts seems to be an aggregate of the selected group. In other words, no matter what topic is selected, it gives you the every exercise attempted in that time frame for that group.

2. Sometimes, when you pick certain topics (IE Addition and subtraction word problems) when the data refreshes after submitting, the repopulating algorithm adds additional, unselected topics (in the case of the previous example, it keeps adding Overview of WW2) (note: that one seems to be a popular choice)
(second note: the data represented does NOT include the added topic, unless you hit submit again)

**Comment by rtibbles** [comment]
@sethachoi looks like we don't have too much to worry about on the memory usage side, thanks to this memory profile by @gwong89 https://gist.github.com/gwong89/7e5f06796f07ac1f33aa

File: kalite/coachreports/api_views.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
 from math import ceil
 import datetime
+import json
+
+import pdb
```
Comment (rtibbles) [inline kalite/coachreports/api_views.py]: Probably want to delete this debugging statement :)

File: kalite/coachreports/api_views.py [hunk 2]
```diff
@@ -179,11 +188,23 @@ def aggregate_learner_logs(request):
...
+            user__in=learners,
+            latest_activity_timestamp__gte=start_date,
+            latest_activity_timestamp__lte=end_date).values_list(id_field, flat=True)
+
+        if len(object_buffer) > 1:
+            all_object_ids.update(object_buffer)
+        elif len(object_buffer) == 1:
+            all_object_ids.add(object_buffer)
+    if len(all_object_ids) > 0:
+        output_dict["content_parents"] = map(lambda x: {"id": x.get("id"), "title": x.get("title")}, get_content_parents(ids=list(all_object_ids)))
```
Comment (rtibbles) [inline kalite/coachreports/api_views.py]: Probably don't need the if statement here, as the map will just give an empty list.

Also should probably map this to "available_topics".

File: kalite/coachreports/api_views.py [hunk 3]
```diff
@@ -135,12 +138,14 @@ def aggregate_learner_logs(request):
...
     end_date = request.GET.get("end_date", None)
 
-    topic_ids = request.GET.getlist("topic_id", [])
+    topic_ids = json.loads(request.GET.get("topic_ids", "[]"))
 
     log_types = request.GET.getlist("log_type", ["exercise", "video", "content"])
 
     output_logs = []
 
+    all_topics = map(lambda x: {"id": x.get("id"), "title": x.get("title")}, get_leafed_topics())
```
Comment (rtibbles) [inline kalite/coachreports/api_views.py]: Can remove this now!

File: kalite/coachreports/api_views.py [hunk 4]
```diff
@@ -149,6 +154,8 @@ def aggregate_learner_logs(request):
         "total_complete": 0,
         "total_struggling": 0,
         "total_not_attempted": 0,
+        "available_topics": all_topics,
```
Comment (rtibbles) [inline kalite/coachreports/api_views.py]: And this!

File: kalite/coachreports/static/js/coachreports/coach_reports/views.js [hunk 5]
```diff
@@ -86,9 +87,48 @@ var CoachSummaryView = BaseView.extend({
...
+        "click #show_tabular_report": "toggle_tabular_view",
+        "click #topic-list-submit": "set_data_model"
     },
 
+
+    /*
+    this function populates the topic selection list
+    */
+    appendTopicList: function() {
+        //var parseData = this.data_model.get("available_topics");
```
Comment (rtibbles) [inline kalite/coachreports/static/js/coachreports/coach_reports/views.js]: I think "available_topics" is more semantic, so would prefer this be changed back.

File: kalite/coachreports/static/js/coachreports/coach_reports/views.js [hunk 6]
```diff
@@ -174,12 +214,14 @@ var CoachSummaryView = BaseView.extend({
         }
 
         if (!this.data_model) {
+            var topic_ids = _.map(this.$("#topic-list option:checked"), function (node) {return node.value;});
+            console.log(topic_ids);
```
Comment (rtibbles) [inline kalite/coachreports/static/js/coachreports/coach_reports/views.js]: https://www.youtube.com/watch?v=RTrAVpK9blw

File: node_modules/bootstrap-multiselect/dist/js/bootstrap-multiselect.js [hunk 7]
```diff
@@ -0,0 +1,1416 @@
+/**
```
Comment (rtibbles) [inline node_modules/bootstrap-multiselect/dist/js/bootstrap-multiselect.js]: This file should not be here (if you have made edits here, they will get overwritten when other people run ```npm install```.

If you have edits, let's sit down to monkey patch in a better way!

**Comment by rtibbles** [comment]
You need to commit your changes to package.json - bootstrap-multiselect is still not showing up in the package.json here.

**Comment by rtibbles** [comment]
Test failures extant on develop. This is good to go. Merging.

**Pull request merged**